src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -135,34 +152,41 @@ FrozenObjectSegment::FrozenObjectSegment(size_t sizeHint) :
         ThrowOutOfMemory();
     }
 
+    m_pStart = static_cast<uint8_t*>(committedAlloc);
+    m_pCurrent = m_pStart + sizeof(ObjHeader);
+    m_SizeCommitted = FOH_COMMIT_SIZE;
+    INDEBUG(m_ObjectsCount = 0);
+
     // ClrVirtualAlloc is expected to be PageSize-aligned so we can expect
     // DATA_ALIGNMENT alignment as well
     _ASSERT(IS_ALIGNED(committedAlloc, DATA_ALIGNMENT));
+}
+
+void FrozenObjectSegment::Register()
+{
+    // Caller is expected to make sure it's not registered twice
+    _ASSERT(!IsRegistered());
 
     segment_info si;
-    si.pvMem = committedAlloc;
+    si.pvMem = m_pStart;
     si.ibFirstObject = sizeof(ObjHeader);
-    si.ibAllocated = si.ibFirstObject;
-    si.ibCommit = FOH_COMMIT_SIZE;
+    si.ibAllocated = (size_t)m_pCurrent; // there can be multiple objects already allocated
+    si.ibCommit = m_SizeCommitted;
     si.ibReserved = m_Size;
 
+    // NOTE: RegisterFrozenSegment may take a GC lock inside.
     m_SegmentHandle = GCHeapUtilities::GetGCHeap()->RegisterFrozenSegment(&si);
     if (m_SegmentHandle == nullptr)
     {
-        ClrVirtualFree(alloc, 0, MEM_RELEASE);
+        ClrVirtualFree(m_pStart, 0, MEM_RELEASE);
```

Comment:
We do not want to free the memory here. We should leave it reserved and assume that some other thread tries to register the segment, it is going to succeed.

(We will leak the frozen object allocation in this situation. I think it is ok.)

Reply:
Do you mean we shouldn't also ThrowOOM here? And we'll just return an object from FrozenObjectManager and pretend everything is ok (let's say there are no other thread, it's just our thread failed to reserve it)

Reply:
We should ThrowOOM here. I do not think we can proceed without the segment being registered with the GC.

Reply:
My guess that generally GC just ignores ro segments except:
1) scanning in-range ro-segements in segments-based GC (e.g. 32bit)
2) reporting them to debugger via EnumMemory

so probably it does need a lock

Reply:
Ok, remove ClrVirtualFree